pulsar-metadata/src/main/java/org/apache/pulsar/metadata/bookkeeper/PulsarRegistrationClient.java

```diff
@@ -153,42 +176,56 @@ public void unwatchReadOnlyBookies(RegistrationListener registrationListener) {
         readOnlyBookiesWatchers.remove(registrationListener);
     }
 
-    private void handleDeletedBookieNode(Notification n) {
-        if (n.getType() == NotificationType.Deleted) {
-            BookieId bookieId = stripBookieIdFromPath(n.getPath());
-            if (bookieId != null) {
-                log.info("Bookie {} disappeared", bookieId);
-                bookieServiceInfoCache.remove(bookieId);
-            }
-        }
-    }
-
-    private void handleUpdatedBookieNode(Notification n) {
-        BookieId bookieId = stripBookieIdFromPath(n.getPath());
-        if (bookieId != null) {
-            log.info("Bookie {} info updated", bookieId);
-            readBookieServiceInfoAsync(bookieId);
-        }
-    }
-
+    /**
+     * This method will receive metadata store notifications and then update the
+     * local cache in background sequentially.
+     */
     private void updatedBookies(Notification n) {
-        if (n.getType() == NotificationType.Created || n.getType() == NotificationType.Deleted) {
-            if (n.getPath().startsWith(bookieReadonlyRegistrationPath)) {
-                getReadOnlyBookies().thenAccept(bookies -> {
-                    readOnlyBookiesWatchers.forEach(w -> executor.execute(() -> w.onBookiesChanged(bookies)));
-                });
-                handleDeletedBookieNode(n);
-            } else if (n.getPath().startsWith(bookieRegistrationPath)) {
-                  getWritableBookies().thenAccept(bookies ->
-                        writableBookiesWatchers.forEach(w -> executor.execute(() -> w.onBookiesChanged(bookies))));
-                handleDeletedBookieNode(n);
-            }
-        } else if (n.getType() == NotificationType.Modified) {
-            if (n.getPath().startsWith(bookieReadonlyRegistrationPath)
-                || n.getPath().startsWith(bookieRegistrationPath)) {
-                handleUpdatedBookieNode(n);
-            }
+        // make the notification callback run sequential in background.
+        final String path = n.getPath();
+        if (!path.startsWith(bookieReadonlyRegistrationPath) || !path.startsWith(bookieRegistrationPath)) {
+            // ignore unknown path
+            return;
         }
+        final BookieId bookieId = stripBookieIdFromPath(n.getPath());
+        log.info("Bookie {} do {}. path: {}", bookieId, n.getType(), n.getPath());
```

Comment:
we should log this only if bookieId is available

Reply:
fixed.